coding/traffic.hpp

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include "coding/reader.hpp"
+#include "coding/writer.hpp"
+
+#include "std/vector.hpp"
+
+// todo(@m) Endianness-agnostic?
+
+namespace coding
+{
+struct TrafficDataPoint
+{
+  TrafficDataPoint() = default;
+
+  TrafficDataPoint(uint64_t timestamp, double lat, double lon): m_timestamp(timestamp), m_lat(lat), m_lon(lon) {}
+
+  // bool operator<(TrafficDataPoint const & o) { return m_timestamp < o.m_timestamp; }
+  
+  uint64_t m_timestamp = 0;
+  double m_lat = 0.0;
+  double m_lon = 0.0;
+};
+
+static double const minLat = -90;
+static double const maxLat = 90;
+static double const minLon = -180;
+static double const maxLon = 180;
+
+uint32_t DoubleToUnsigned(double x, double min, double max, uint32_t bits);
+
+double UnsignedToDouble(uint32_t x, double min, double max, uint32_t bits);
+
+template <typename Writer>
+void SerializeDataPoints(Writer & writer, vector<TrafficDataPoint> const & points)
+{
+  uint32_t const coordBits = 30;
+
+  // todo(@m) Rewrite to varints?
+  
+  auto const startPos = writer.Pos();
+  uint32_t header = 0;
+  // We will fill it later.
+  writer.Write(&header, sizeof(header));
+  
+  if (points.empty())
+    return;
+
+  uint64_t const firstTimestamp = points[0].m_timestamp;
+  uint32_t const firstLat = DoubleToUnsigned(points[0].m_lat, minLat, maxLat, coordBits);
+  uint32_t const firstLon = DoubleToUnsigned(points[0].m_lon, minLon, maxLon, coordBits);
+  writer.Write(&firstTimestamp, sizeof(uint64_t));
+  writer.Write(&firstLat, sizeof(uint32_t));
+  writer.Write(&firstLon, sizeof(uint32_t));
+
+  for (size_t i = 1; i < points.size(); ++i)
+  {
+    ASSERT_LESS_OR_EQUAL(points[i - 1].m_timestamp, points[i].m_timestamp, ());
+    
+    uint64_t const deltaTimestamp = points[i].m_timestamp - points[i - 1].m_timestamp;
+    // todo(@m) Adjust min/max bounds here and write a test for it.
+    // and write a test for coordbits
+    uint32_t deltaLat = DoubleToUnsigned(points[i].m_lat - points[i - 1].m_lat, minLat, maxLat, coordBits);
+    uint32_t deltaLon = DoubleToUnsigned(points[i].m_lon - points[i - 1].m_lon, minLon, maxLon, coordBits);
+    
+    writer.Write(&deltaTimestamp, sizeof(uint64_t));
+    writer.Write(&deltaLat, sizeof(uint32_t));
+    writer.Write(&deltaLon, sizeof(uint32_t));
+  }
+  
+  auto const endPos = writer.Pos();
+  // header = static_cast<uint32_t>(endPos - startPos);
+  header = points.size();
+  writer.Seek(startPos);
```

Comment:
Usually, socket does not allow you to seek. Why not to write number of points as var-int instead?